### ISSUE TITLE
Gdx.graphics.getDensity() returns infinity inside VirtualBox

### ISSUE THREAD

**Issue by ichernev** [libgdx/libgdx#6537]
#### Issue details
Calling `Gdx.graphics.getDensity()` inside VirtualBox (Linux guest/host) returns `Infinity` (because display size reported by Lwjgl is 0). I understand that on some platforms the display density can not be computed, but I think there should be a special value UNKNOWN or an exception thrown to force client code to handle this case. I got an exception in the JNI code of FreeType (because I use density for font size scaling), and getDensity() doesn't indicate it can return a completely invalid value.

It returns an estimate in any case, so if the value is too low / too high it can just default to 1 or something.

#### Version of LibGDX and/or relevant dependencies
GDX version `1.10.0` (latest at time of writing)

#### Please select the affected platforms
- [ ] Android
- [ ] iOS
- [ ] HTML/GWT
- [ ] Windows
- [x] Linux
- [ ] MacOS

**Comment by crykn** [comment]
Could you please try whether the issue also happens with the [LWJGL 3 backend](https://gist.github.com/crykn/eb37cb4f7a03d006b3a0ecad27292a2d)?

**Comment by ichernev** [comment]
This is with Lwjgl3. It is not trivial to switch to 2 at this point :)

**Comment by tommyettinger** [comment]
Good point that it already returns an estimate; I think you're right in your interpretation of the docs that it does not indicate it can return a non-finite value, and an estimate shouldn't generally be infinite or NaN. A density of 0 would clearly indicate that something is wrong with the estimation, since there's no way you could have a screen with 0 pixels per inch/cm that actually works. Same for a negative number. I'm not sure if it's better to just return 1, indicating "some normal screen," if the screen is actually unusable at that point in time. A quick workaround might be to use Float.isFinite() as your check for an invalid result, since that would pick up both infinities and also NaN (and NaN is tricky to check for normally). I don't know if getDensity() can return NaN, but if it's dividing a positive number by 0.0f already to get Infinity, it's not unreasonable that it could divide 0 by 0.0f and get NaN.

**Comment by crykn** [comment]
I think you should also report this to the [GLFW project](https://github.com/glfw/glfw), since the bug seems to be [on their end](https://github.com/libgdx/libgdx/blob/add6031b113d9e255f791656581bdffbbff83f89/backends/gdx-backend-lwjgl3/src/com/badlogic/gdx/backends/lwjgl3/Lwjgl3Graphics.java#L253).

**Comment by Timvisser1994** [comment]
Hi there, the following is a possible fix:
```
AbstractGraphics.cs lines 10-20

/**
* Facade method that estimates the pixel density of the underlying Graphics implementation.
* @return The approximated pixel density, or 1 if this could not be determined
*/
@Override
public float getDensity () {
	float ppiX = getPpiX();
	// Guard against invalid values
	if (ppiX <= 0 || Float.isNaN(ppiX)) return 1f;
	return ppiX / 160f;
}
```

The solution posted by @tommyettinger sounds good here, but unfortunately, `Float.isFinite()` is only available from Java 1.8 and up. I did not have the rights to push to a new branch, and open up a Pull Request. So here you go 😊

**Comment by tommyettinger** [comment]
@Timvisser1994 That won't help with the original problem, which is a density of Infinity. Infinity wouldn't be considered `<= 0` and `isNaN()` returns false. NaN is an oddity, but looking at the OpenJDK source code gives a solution:
```java
public float getDensity () {
	float ppiX = getPpiX();
	return (ppiX > 0 && ppiX <= Float.MAX_VALUE) ? ppiX / 160f : 1f;
}
```
The important thing here is that all valid finite floats are `<= Float.MAX_VALUE` (NaN is not, so it gets ruled out), and we only want positive and non-zero floats, so the check for `> 0` also rules out negative infinity. The check against MAX_VALUE  is used by isFinite() in Java 8. I changed `getDensity()` to single-return form because I think some JDKs can analyze that form a little easier, or at least it won't be harder.

There's still the issue that an invalid PPI is now indistinguishable from a common one, `1f`. I think in the cases where we would get an invalid PPI, the display density doesn't really matter as long as it doesn't cause problems, so `1f` is probably fine; users could check on their own if `getPpiX()` returns a value they can handle, also.

**Comment by Timvisser1994** [comment]
@tommyettinger Great, indeed this improves it. I don't think that the indistinguisibility is bad, because `getDensity()` returns an approximate anyways. To me, `1f` (one pixel per unit) is true always, except when the screen is better than that.

Anyways, that's my two cents 🙂

**Comment by tommyettinger** [comment]
That's true, it is approximate, so 1f as a general guess is probably fine. It could maybe do something like 0.015625f density as an indicator of something very strange happening, but still a technically possible resolution. That's 1/64, which would correspond to 2.5 pixels per inch or less than 1 pixel per centimeter, and that seems unlikely to appear. Maybe a monome, or some other similar hardware, would have that approximate resolution, but I doubt one of those is a valid display for LWJGL3. But yeah, that's not needed because `getPpiX()` is still available and can be used to check if the backing information is something strange, even if `getDensity()` returns 1f for valid and invalid situations. Thanks for hopping on to the Discord, by the way; if you encounter any issues with adding a PR, feel free to ask in the `contributing` channel there (there are several "gotcha" problems that can affect PRs, chief among them formatting being automatically changed).

**Comment by crykn** [comment]
This has been fixed in #6539, however, it would still be appreciated if you reported this issue (i.e. `glfwGetMonitorPhysicalSize` returns the wrong display size) to the [GLFW project](https://github.com/glfw/glfw) with some information about your specific setup.

**Comment by ichernev** [comment]
@crykn I will report to GLFW. The setup is not too crazy -- virtualbox + linux (Arch). In any case I'm glad you worked around it. There is always a chance that GLFW will return wrong data.
